File: src/org/genyris/io/PrefixMapper.java
// Copyright 2009 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;

import org.genyris.core.PrefixSymbol;
import org.genyris.core.SimpleSymbol;
import org.genyris.core.Symbol;
import org.genyris.exception.GenyrisException;

public class PrefixMapper {
    public static final String ABBREVIATION_SEPARATOR_CHAR = ":";
	private Map _abbreviations; // #TODOuse a Map<>
    private char _dynaChar;

    public PrefixMapper(char dynaChar) {
        _abbreviations = new HashMap();
        _dynaChar = dynaChar;
    }


    public void addAbbreviation(String abbrev, String uri) throws GenyrisException {
    	if(abbrev.equals(ABBREVIATION_SEPARATOR_CHAR)) {
    		abbrev = "";
    	}
        try {
            new URL(uri);
        }
        catch (MalformedURLException e) {
            throw new GenyrisException("namespace for '" + abbrev +"' is not a valid URL: " + uri);
        }
        if(abbrev.startsWith(String.valueOf(_dynaChar))) {
            throw new GenyrisException("cannot start a abbreviation with " + _dynaChar + " in parse: " + abbrev);
        }
        if (_abbreviations.containsKey(abbrev)) {
            if(!_abbreviations.get(abbrev).equals(uri)) {
                throw new GenyrisException("conflicting abbreviation in parse: " + abbrev + " " + uri);
            }
        } else {
            _abbreviations.put(abbrev, uri);
        }
    }

    public static boolean hasAbbreviation(String symbol) {
        return symbol.contains(ABBREVIATION_SEPARATOR_CHAR);
    }

    private static String getAbbreviation(String symbol) {
        return symbol.substring(0, symbol.indexOf(ABBREVIATION_SEPARATOR_CHAR));
    }

    private static String getLocalname(String symbol) {
        return symbol.substring(symbol.indexOf(ABBREVIATION_SEPARATOR_CHAR) + 1);
    }

    public SimpleSymbol symbolFactory(String news) throws GenyrisException {
        if(news.equals(ABBREVIATION_SEPARATOR_CHAR) || !hasAbbreviation(news) ) {
            return Symbol.symbolFactory(news, false);
        }
        else {
            String abbrev = getAbbreviation(news);
            String localName = getLocalname(news);
            if (!_abbreviations.containsKey(abbrev)) {
                throw new GenyrisException("Unknown abbreviation: " + abbrev);
            } else {
                String prefix = (String)_abbreviations.get(abbrev);
                PrefixSymbol pre =  new PrefixSymbol(prefix, localName, abbrev);
                return pre;
            }
        }
    }
    public String getCannonicalSymbol(String news) throws GenyrisException {
        String abbrev;
        if(news.equals(ABBREVIATION_SEPARATOR_CHAR) || !hasAbbreviation(news) ) {
            return news;
        }
        else {
            abbrev = getAbbreviation(news);
            if (!_abbreviations.containsKey(abbrev)) {
                throw new GenyrisException("Unknown prefix: " + abbrev);
            } else {
                return _abbreviations.get(abbrev) + getLocalname(news);
            }
        }
    }

}

File: src/org/genyris/io/PrefixMapper.kt
// Copyright 2009 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.io

import org.genyris.core.PrefixSymbol
import org.genyris.core.SimpleSymbol
import org.genyris.core.Symbol
import org.genyris.exception.GenyrisException
import java.lang.String
import java.net.URI
import java.net.URISyntaxException
import kotlin.Any
import kotlin.Boolean
import kotlin.Char
import kotlin.collections.HashMap
import kotlin.collections.MutableMap
import kotlin.plus
import kotlin.toString

class PrefixMapper(dynaChar: Char) {
    private val _abbreviations: MutableMap<*, *> // #TODOuse a Map<>
    private val _dynaChar: Char

    init {
        _abbreviations = HashMap<Any?, Any?>()
        _dynaChar = dynaChar
    }


    @kotlin.Throws(GenyrisException::class)
    fun addAbbreviation(abbrev: String, uri: String) {
        var abbrev = abbrev
        if (abbrev == ABBREVIATION_SEPARATOR_CHAR) {
            abbrev = ""
        }
        try {
            URI(uri)
        } catch (e: URISyntaxException) {
            throw GenyrisException("namespace for '" + abbrev + "' is not a valid URI: " + uri)
        }
        if (abbrev.startsWith(String.valueOf(_dynaChar))) {
            throw GenyrisException("cannot start a abbreviation with " + _dynaChar + " in parse: " + abbrev)
        }
        if (_abbreviations.containsKey(abbrev)) {
            if (_abbreviations.get(abbrev) != uri) {
                throw GenyrisException("conflicting abbreviation in parse: " + abbrev + " " + uri)
            }
        } else {
            _abbreviations.put(abbrev, uri)
        }
    }

    @kotlin.Throws(GenyrisException::class)
    fun symbolFactory(news: kotlin.String): SimpleSymbol {
        if (news == ABBREVIATION_SEPARATOR_CHAR || !hasAbbreviation(news)) {
            return Symbol.Companion.symbolFactory(news, false)
        } else {
            val abbrev: kotlin.String = getAbbreviation(news)
            val localName: kotlin.String = getLocalname(news)
            if (!_abbreviations.containsKey(abbrev)) {
                throw GenyrisException("Unknown abbreviation: " + abbrev)
            } else {
                val prefix = _abbreviations.get(abbrev) as kotlin.String?
                val pre = PrefixSymbol(prefix, localName, abbrev)
                return pre
            }
        }
    }

    @kotlin.Throws(GenyrisException::class)
    fun getCannonicalSymbol(news: kotlin.String): kotlin.String {
        val abbrev: kotlin.String?
        if (news == ABBREVIATION_SEPARATOR_CHAR || !hasAbbreviation(news)) {
            return news
        } else {
            abbrev = getAbbreviation(news)
            if (!_abbreviations.containsKey(abbrev)) {
                throw GenyrisException("Unknown prefix: " + abbrev)
            } else {
                return _abbreviations.get(abbrev).toString() + getLocalname(news)
            }
        }
    }

    companion object {
        const val ABBREVIATION_SEPARATOR_CHAR: kotlin.String = ":"
        fun hasAbbreviation(symbol: kotlin.String): Boolean {
            return symbol.contains(ABBREVIATION_SEPARATOR_CHAR)
        }

        private fun getAbbreviation(symbol: kotlin.String): kotlin.String {
            return symbol.substring(0, symbol.indexOf(ABBREVIATION_SEPARATOR_CHAR))
        }

        private fun getLocalname(symbol: kotlin.String): kotlin.String {
            return symbol.substring(symbol.indexOf(ABBREVIATION_SEPARATOR_CHAR) + 1)
        }
    }
}
